### PR TITLE
feat (verifier): add risc_zero verifier ffi

### LINKED ISSUE #169: feat (verifier): RISC Zero verifier ffi
# Description
- Implement RISC Zero verifier ffi.

### PR DESCRIPTION
**Pull request by NicolasRampoldi**
# Description
- This PR adds the Risc0 verifier FFI.
- To test, run:
  - `make generate_risc_zero_fibonacci_proof`. You will see `error: invalid inline table` but this is a problem of the risc0 library, the program will compile anyways.
  - Then, run `make test_risc_zero_rust_ffi`. To run the rust tests.
  - Finally, run `make test_risc_zero_go_bindings_macos` or `make test_risc_zero_go_bindings_linux` depending on your OS to test the go bindings.

### CODE REVIEW

File: operator/risc_zero/lib/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        assert!(result)
+    }
+
+    #[test]
+    fn verify_risc_zero_aborts_with_bad_proof() {
+        const RECEIPT_SIZE: usize = RECEIPT.len();
+        let mut receipt_buffer = [42u8; super::MAX_RECEIPT_SIZE];
+        receipt_buffer[..RECEIPT_SIZE].clone_from_slice(RECEIPT);
+
+        // Leer el IMAGE_ID desde el archivo
```
Comment (taturosati) [inline operator/risc_zero/lib/src/lib.rs]: Comment should be in English

File: operator/risc_zero/lib/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        Ok(image_id)
+    }
+
+    #[test]
+    fn verify_risc_zero_receipt_with_image_id_works() {
+        const RECEIPT_SIZE: usize = RECEIPT.len();
+        let mut receipt_buffer = [0u8; super::MAX_RECEIPT_SIZE];
+        receipt_buffer[..RECEIPT_SIZE].clone_from_slice(RECEIPT);
+
+        // Leer el IMAGE_ID desde el archivo
```
Comment (taturosati) [inline operator/risc_zero/lib/src/lib.rs]: Comment should be in English

File: operator/risc_zero/lib/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+[package]
+name = "risc_zero_verifier_ffi"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+risc0-zkvm = { git = "https://github.com/risc0/risc0",  rev = "6107a65aa20816be9c62da0fa507794ebb1a682e" }
```
Comment (entropidelic) [inline operator/risc_zero/lib/Cargo.toml:7]: we should create an issue or some reminder for us to check when the next stable release is available

File: operator/risc_zero/lib/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
+use risc0_zkvm::Receipt;
+
+pub const MAX_RECEIPT_SIZE: usize = 215523;
+pub const MAX_IMAGE_ID_SIZE: usize = 8;
```
Comment (entropidelic) [inline operator/risc_zero/lib/src/lib.rs]: this value is fixed, right? so no need to make a constant for it

File: operator/risc_zero/lib/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+#[no_mangle]
+pub extern "C" fn verify_risc_zero_receipt_ffi(
+    receipt_bytes: &[u8; MAX_RECEIPT_SIZE],
+    receipt_len: usize,
+    image_id: &[u32; MAX_IMAGE_ID_SIZE],
```
Comment (entropidelic) [inline operator/risc_zero/lib/src/lib.rs]:
```suggestion
    image_id: &[u32; 8],
```

File: operator/risc_zero/risc_zero.go [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+#cgo darwin LDFLAGS: ${SRCDIR}/lib/librisc_zero_verifier_ffi.dylib
+
+#include "lib/risc_zero.h"
+*/
+import "C"
+import "unsafe"
+
+const (
+	MaxReceiptSize = 215523
+	MaxImageIdSize = 8
```
Comment (entropidelic) [inline operator/risc_zero/risc_zero.go]: as with the rust part of the ffi, there is no need to specify this constant here

**Comment by entropidelic** [comment]
looks good to me, just a minor comment on the `MaxImageIdSize` constant in the Go FFI code which is not needed.

Once the risc0 verifier is integrated in the full flow, the readme should be updated with the steps to generate and verify the proofs.

**Review by entropidelic — APPROVED** [review]

**Review by taturosati — COMMENTED** [review]
lgtm

File: operator/risc_zero/risc_zero.go (hunk 6 above)
Comment (NicolasRampoldi) [reply operator/risc_zero/risc_zero.go]: Fixed in [d4841c1](https://github.com/yetanotherco/aligned_layer/pull/171/commits/d4841c1c592c50e1b3c692a8162b88b0e9e61da1)

File: operator/risc_zero/lib/src/lib.rs (hunk 1 above)
Comment (NicolasRampoldi) [reply operator/risc_zero/lib/src/lib.rs]: That's from an old commit

File: operator/risc_zero/lib/src/lib.rs (hunk 2 above)
Comment (NicolasRampoldi) [reply operator/risc_zero/lib/src/lib.rs]: That's from an old commit

**Review by taturosati — APPROVED** [review]

**Pull request merged**